### PR TITLE
Updates for PyTorch v1.7.1

### LINKED ISSUE #5: Updates for PyTorch v1.7.1
## Issue description

Nimble was initially developed on top of PyTorch v1.4, which was the latest version at the time of the development.
Now that PyTorch went through three major-version updates from v1.4.1 and started to support CUDA 11, we upgrade Nimble to be compatible with PyTorch v1.7.1 and CUDA 11.0. (+ torchvision 0.8)
New installation instructions can be found [here](https://github.com/WoosukKwon/nimble/blob/nimble-v1.7.1/NIMBLE_INSTALL.md).

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR closes #5.

### CODE REVIEW

File: aten/src/THCUNN/generic/SpatialConvolutionMM.cu [hunk 1]
```diff
@@ -158,7 +158,9 @@ void THNN_(SpatialConvolutionMM_updateOutput)(
   THCTensor_(resize4d)(state, output, batchSize, nOutputPlane, outputHeight, outputWidth);
 
   // Resize temporary columns
-  THCTensor_(resize2d)(state, columns, nInputPlane*kW*kH, outputHeight*outputWidth);
+  if (kW != 1 || kH != 1) {
```
Comment (gyeongin) [inline aten/src/THCUNN/generic/SpatialConvolutionMM.cu]: Looks like PyTorch v1.7.1 has a bug here.
If we run
```python
import torch
torch.manual_seed(0)
m = torch.nn.Conv2d(2,3,kernel_size=1,padding=1).cuda()
x = torch.randn(1,2,5,5).cuda()
print(m(x)) # cudnn
torch.backends.cudnn.enabled=False
print(m(x))
```
The output is different between the two runs.
I think we should closely follow `main_pytorch_v1.4.1` branch's implementation.

Comment (gyeongin) [reply aten/src/THCUNN/generic/SpatialConvolutionMM.cu]: Also we can skip line 171's `THCTensor_(fill)` operation if we do not need bias.
This operation will add an additional `memset` GPU task.
Similarly, we can skip line 212's `THCTensor_(zero)` operation (refer to `main_pytorch_v1.4.1` branch's implementation).

File: torch/jit/_trace.py [hunk 2]
```diff
@@ -1046,7 +1046,7 @@ def check_unique(param):
 
         self.__dict__["_name"] = type(orig).__name__
         self.__dict__["_actual_script_module"] = script_module
-        for name in ("_parameters", "_buffers", "_modules"):
+        for name in ("_parameters", "_buffers", "_modules", "training"):
```
Comment (gyeongin) [inline torch/jit/_trace.py:1049]: Can you add a comment for why we need this?

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -719,7 +722,28 @@ void initJITBindings(PyObject* module) {
...
+      })
+      .def("_jit_required_passes",
+          [](std::shared_ptr<Graph>& graph) {
+            Inline(*graph);
+            LowerSimpleTuples(graph);
+            ConstantPooling(graph);
+            runRequiredPasses(graph);
+            ConstantPropagation(graph);
+            EliminateDeadCode(graph);
+            EliminateCommonSubexpression(graph);
```
Comment (gyeongin) [inline torch/csrc/jit/python/init.cpp:738]: How did you select these passes?
It would be great if you point out which code is relevant to this decision.

**Review by gyeongin — COMMENTED** [review]
Thanks for your effort on this! The change looks good.
I've left a few minor comments.

File: torch/jit/_trace.py (hunk 2 above)
Comment (WoosukKwon) [reply torch/jit/_trace.py:1049]: PyTorch v1.7.1 has a small bug in setting the `training` attribute for traced Module.
This bug is fixed in the master branch of PyTorch.
Please refer to [this](https://github.com/pytorch/pytorch/commit/ecfa7a27b8e6d121456c949688f2c01c6a6b7300).

File: aten/src/THCUNN/generic/SpatialConvolutionMM.cu (hunk 1 above)
Comment (WoosukKwon) [reply aten/src/THCUNN/generic/SpatialConvolutionMM.cu]: I've resolved the bug and inefficiencies as you have pointed out. Please check again.

File: torch/csrc/jit/python/init.cpp (hunk 3 above)
Comment (WoosukKwon) [reply torch/csrc/jit/python/init.cpp:738]: I added a few more compilation passes and left some comments about where the code came from. Please check.

File: aten/src/THCUNN/generic/SpatialConvolutionMM.cu (hunk 1 above)
Comment (gyeongin) [reply aten/src/THCUNN/generic/SpatialConvolutionMM.cu:179]: I think you've forgot to remote line 178-190 by mistake. Let's remote these lines.

File: aten/src/THCUNN/generic/SpatialConvolutionMM.cu [hunk 4]
```diff
@@ -157,8 +157,28 @@ void THNN_(SpatialConvolutionMM_updateOutput)(
...
+  if (bias) {
+    // Define a buffer of ones, for bias accumulation
+    // Note: this buffer can be shared with other modules, it only ever gets increased,
+    // and always contains ones.
+    if (ones->dim() != 2 || ones->size(0)*ones->size(1) < outputHeight*outputWidth) {
+      // Resize plane and fill with ones...
+      THCTensor_(resize2d)(state, ones, outputHeight, outputWidth);
+      THCTensor_(fill)(state, ones, ScalarConvert<int, scalar_t>::to(1));
+    }
+  }
```
Comment (gyeongin) [inline aten/src/THCUNN/generic/SpatialConvolutionMM.cu:176]: I think you've forgot to remote line 178-190 by mistake. Let's remote these lines.

**Comment by peri044** [comment]
Hello @WoosukKwon, I've cloned your [branch ](https://github.com/WoosukKwon/nimble/tree/nimble-v1.7.1)and followed your instructions [https://github.com/WoosukKwon/nimble/blob/nimble-v1.7.1/NIMBLE_INSTALL.md](url) on building nimble with Pytorch 1.7.1. 
I face a CMakeError related to pthread. Here is the complete error log [https://gist.github.com/peri044/f110aab43a35de056aa752c10cad051d](url) 
Can you please advise on what might be the issue here ? Thanks !!

**Comment by WoosukKwon** [comment]
Hello @peri044, could you try building PyTorch v1.7.1 without Nimble? Checkout to branch `torch-v1.7.1` and follow the instructions. The error seems unrelated to Nimble. It's highly likely that the error comes from AVX or other dependencies of PyTorch.

File: aten/src/THCUNN/generic/SpatialConvolutionMM.cu (hunk 4 above)
Comment (WoosukKwon) [reply aten/src/THCUNN/generic/SpatialConvolutionMM.cu:176]: My bad! Thanks for letting me know! I've just removed the lines. Please check again.

**Comment by peri044** [comment]
@WoosukKwon Thanks for your quick response

1) As per your comment, I tried building torch-v1.7.1 branch and faced the exact same error ([CMakeError](https://gist.github.com/peri044/f110aab43a35de056aa752c10cad051d))

2) I also tried building nimble using pytorch 1.4 and cuda-10.2 (default config) from [https://github.com/snuspl/nimble/blob/main/NIMBLE_INSTALL.md](url)
The build was successful. However I see import torch error (tried outside nimble directory as well) as follows 
```
import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dperi/Downloads/anaconda3/envs/nimble/lib/python3.7/site-packages/torch/__init__.py", line 81, in <module>
    from torch._C import *
ImportError: /home/dperi/Downloads/anaconda3/envs/nimble/lib/python3.7/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch8autograd6Engine11thread_mainERKSt10shared_ptrINS0_9GraphTaskEEb

```
Any suggestions here ? Thanks

**Comment by WoosukKwon** [comment]
@peri044 Sorry for the late response.
Do you still have the same problem? I think you have a trouble in building PyTorch from source.
I've never seen exactly the same error message so far, but in my experience, similar build errors were removed by cleaning up the old versions of PyTorch (e.g., repeat `pip uninstall torch` more than twice) and re-installing the new version.
If the error persists, please report to PyTorch.
Thanks.

**Comment by gyeongin** [comment]
Looks good to merge.

**Comment by bgchun** [comment]
@WoosukKwon Could you update the PyTorch version Nimble supports in README and INSTALLATION?

**Pull request merged**